contrib/epee/include/string_tools.h

```diff
@@ -105,19 +105,23 @@ namespace string_tools
       return false;
     try
     {
-      long v = 0;
-      for(size_t i = 0; i < (s.size() + 1) / 2; i++)
+      res.resize((s.size() + 1) / 2);
+      for (size_t i = 0; i < s.size(); i += 2)
       {
-        CharT byte_str[3];
-        size_t copied = s.copy(byte_str, 2, 2 * i);
-        byte_str[copied] = CharT(0);
-        CharT* endptr;
-        v = strtoul(byte_str, &endptr, 16);
-        if (v < 0 || 0xFF < v || endptr != byte_str + copied)
+        int tmp = s[i] - '0';
+        if (tmp < 0) return false;
+        if (tmp > 9) tmp = (tmp & 0x17) - 7;
+        if (tmp > 15) return false;
+        tmp <<= 4;
+        if (s[i+1])
```

Comment:
This could be an invalid memory access here (and below) if `allow_partial_byte = true`. Might be able to drop that parameter though.

Reply:
git grep doesn't turn up any invocations where that param is set. We should probably just remove it and check for an odd size at the top, and immediately fail.

Reply:
````
 template<class CharT>
  bool parse_hexstr_to_binbuff4(const std::basic_string<CharT>& s, std::basic_string<CharT>& res)
  {
    res.clear();
    if (s.size() & 1)
      return false;
    try
    {
      res.resize(s.size() / 2);
      unsigned char *dst = (unsigned char *)res.data();
      unsigned char *src = (unsigned char *)s.data();
      for(size_t i = 0; i < s.size(); i += 2)
      {
  int tmp = *src++ - '0';
  if (tmp < 0) return false;
  if (tmp > 9) tmp = (tmp & 0x17) - 7;
  if (tmp > 15) return false;
  tmp <<= 4;
    int t2 = *src++ - '0';
    if (t2 < 0) return false;
    if (t2 > 9) t2 = (t2 & 0x17) - 7;
    if (t2 > 15) return false;
    *dst++ = tmp | t2;
      }

      return true;
    }catch(...)
    {
      return false;
    }
}
````
A couple percent faster here.